src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,450 @@
+        idxInChild = childGenome.GetNeuronIndex(momGenome.aLinkGenes[i].ToNeuronId());
+        idxInParent = momGenome.GetNeuronIndex(momGenome.aLinkGenes[i].ToNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(momGenome.aNeuronGenes[idxInParent]);
+        }
+
+        continue;
+      }
+
+      if (idx != -1 && momGenome.aLinkGenes[i].Enabled() && mlpack::math::Random() >= 0.5) {
+        childGenome.AddLink(dadGenome.aLinkGenes[idx]);
+
+        // Add from neuron   TODO: make it a function?? check whether crossover is correct.
+        ssize_t idxInChild = childGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].FromNeuronId());
+        ssize_t idxInParent = dadGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].FromNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(dadGenome.aNeuronGenes[idxInParent]);
+        }
+
+        // Add to neuron
+        idxInChild = childGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].ToNeuronId());
+        idxInParent = dadGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].ToNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(dadGenome.aNeuronGenes[idxInParent]);
+        }
+
+        continue;
+      }
+    }
+  }
+
+  void Crossover(Genome& genome1, Genome& genome2, Genome& childGenome) {
+    if (Species::CompareGenome(genome1, genome2)) {  // genome1 is better
+      CrossoverLinkAndNeuron(genome1, genome2, childGenome);
+    } else {
+      CrossoverLinkAndNeuron(genome2, genome1, childGenome);
+    }
+  }
+
+  // Measure two genomes' disjoint (including exceed).
+  // TODO: we can seperate into disjoint and exceed.
+  // But currently maybe it is enough.
+  double Disjoint(Genome& genome1, Genome& genome2) {
+    double numDisjoint = 0;
+
+    for (ssize_t i=0; i<genome1.NumLink(); ++i) {
+      ssize_t innovId = genome1.aLinkGenes[i].InnovationId();
+      ssize_t idx = genome2.GetLinkIndex(innovId);
+      if (idx == -1 && genome1.aLinkGenes[i].Enabled()) {
+        ++numDisjoint;
+      }
+    }
+
+    for (ssize_t i=0; i<genome2.NumLink(); ++i) {
+      ssize_t innovId = genome2.aLinkGenes[i].InnovationId();
+      ssize_t idx = genome1.GetLinkIndex(innovId);
+      if (idx == -1 && genome2.aLinkGenes[i].Enabled()) {
+        ++numDisjoint;
+      }
+    }
+
+    ssize_t largerGenomeSize = std::max(genome1.NumLink(), genome2.NumLink());
+    double deltaD = numDisjoint / largerGenomeSize;
+    return deltaD; 
+  }
+
+  // Measure two genomes' weight difference.
+  double WeightDiff(Genome& genome1, Genome& genome2) {
+    double deltaW = 0;
+    ssize_t coincident = 0;
+
+    for (ssize_t i=0; i<genome1.NumLink(); ++i) {
+      ssize_t innovId = genome1.aLinkGenes[i].InnovationId();
+      ssize_t idx = genome2.GetLinkIndex(innovId);
+      if (idx != -1 && genome1.aLinkGenes[i].Enabled()) {
+        deltaW += std::abs(genome1.aLinkGenes[i].Weight() - genome2.aLinkGenes[idx].Weight());
```

Comment:
According to the NEAT paper, only calculate the difference of matched links. 
Also, I just found that crossover and distance calculation also consider disabled links. Thus, I revised corresponding functions.